docs: :memo: major rewrite of the naming scheme to match frictionless

lwjohnst86 opened the pull request:
## Description

These changes are done to start converting over to frictionless standard language, since we decided to use frictionless standard. This also really simplifies things. I also rearranged to have a function name section and an app URL section, since they look a bit differently.

## Reviewer Focus

<!-- Select quick/in-depth as necessary -->
This PR needs a bit of an in-depth review.

lwjohnst86 committed:
docs: :memo: major rewrite of the naming scheme to match frictionless

signekb commented on docs/design/naming.qmd:
```diff
@@ -5,74 +5,137 @@ title: "Naming scheme"
-# Update or delete metadata of a specific project
-projects <id> metadata <id> update
-projects <id> metadata <id> delete
+`packages/<id>/resources/<id>/add`
```
So how we differentiate between what we before called data and metadata isn't necessary anymore in the urls since it's only about data when we `add` and otherwise it's about the metadata?

lwjohnst86 replied:
I'm not sure I understand. What do we need to differentiate on?

signekb commented on docs/design/naming.qmd:
```diff
@@ -81,45 +144,46 @@ environment with Postgres installed, the original input from the user
--   Types of folders: raw, data
+-   Types of files: JSON, Parquet, SQLite, compressed raw data (if
+    appropriate).
+-   Types of folders: raw and resources.
```
But maybe Django has a way to handle this?

lwjohnst86 replied:
I will clarify this, the function names are for core, not the app. The app will likely store everything in a Postgres database.

signekb reviewed (commented):
Very nice. I think this is getting a lot clearer. I do have some questions though - specifically about how we want to store raw data.

lwjohnst86 commented on docs/design/naming.qmd:
```diff
@@ -5,74 +5,137 @@ title: "Naming scheme"
--   Data objects do not have any `view` action, since we want to limit
-    access to looking at raw data.
+-   Database and resource file objects do not have any `view` action, as
+    we want to limit access viewing data in the raw files and database.
```
Yup! And it is also out of scope of Sprout.

lwjohnst86 replied:
If we provide the option to view data, we increase the risk that someone else can get the data. We have to think about more things, more security aspects, about permissions, about how they are given, about documentation to inform the user that data contributors can see the data too. So I'd rather avoid that and instead create better reporting summaries aimed at the admin. At the end of the day, the admin user will always be able to see the data anyway, so this isn't super important for us to spend so much time on.

signekb replied:
Alright. Thanks for elaborating!

martonvago reviewed (commented):
Looks great and makes sense to me!

signekb commented on docs/design/naming.qmd:
```diff
@@ -81,45 +144,46 @@ environment with Postgres installed, the original input from the user
+
+`ROOT/packages/<id>/resources/<id>/raw/<timestamp>-<uuid>.<extension>.gzip`
+
+:   Compress all raw data files.
```
```suggestion
:   Compressed raw data files associated with a specific resource.
```
"Compress" makes it sound like an action and not a path :)

signekb commented on docs/design/naming.qmd:
```diff
@@ -81,45 +144,46 @@ environment with Postgres installed, the original input from the user
+
+:   Compress all raw data files.
+
+`ROOT/packages/<id>/resources/<id>/data.parquet`
```
Will the file(s) actually be called "data" or is it a placeholder like <id>? Bc there might be multiple files associated with a resource, right? Or will the addition of a data file (a raw) be added to the one parquet file?
If a placeholder, should we then add `<>` around data?
```suggestion
`ROOT/packages/<id>/resources/<id>/<data>.parquet`
```

lwjohnst86 replied:
It is not a placeholder. Each data resource will only have one data table. But there will be potentially many raw data files, which will be merged into one Parquet table. The name `data` is arbitrary, it could be anything, but let's just keep it simple 😛

signekb commented on docs/design/naming.qmd:
```diff
@@ -81,45 +144,46 @@ environment with Postgres installed, the original input from the user
+    to act as a backup to the database and to act as a faster file
+    format for potential uses in analytics.
+
+`ROOT/packages/<id>/database.sqlite`
```
Same as above. Will it actually be called "database"? If we only have one db per project this is probably fine?

lwjohnst86 replied:
That's right! Only one database per package.

signekb reviewed (commented):
Amazing! 🔥 This is super helpful - and useful! I have added some questions and suggestions.

lwjohnst86 commented on docs/design/naming.qmd:
```diff
@@ -5,74 +5,137 @@ title: "Naming scheme"
+    packages and resources.
 
-Additional "rules" to the naming include:
+### Functions
```
I realize from the comments and also looking through both PRs that this doc is deviating from simply a naming scheme and more into a design. So I think I will split it up into two new PRs that describe the core functionality in one and the app URLs/pages in another.

signekb replied:
Sounds amazing 🎉

lwjohnst86 commented on docs/design/naming.qmd:
```diff
@@ -5,74 +5,137 @@ title: "Naming scheme"
+
+:   Delete a specific package.
+
+`package_resources_list(package_id)`
```
Yes, that's intentional. Listing is many. But the way the URLs are structured, it implies many, unless a specific ID is given.

lwjohnst86 commented on docs/design/naming.qmd:
```diff
@@ -5,74 +5,137 @@ title: "Naming scheme"
+
+:   Create a new resource in a specific package using data and the properties metadata.
+
+`package_resource_add(package_id, resource_id, data)`
+
+:   Add more data to a resource of a specific package.
```
No. But I'll try to expand on this in another PR, based on my comment above.

signekb replied:
Amazing, thanks! As I understand it, from your comments in the two current PRs, we'll have multiple raw files per resource but only one parquet?

lwjohnst86 replied:
See #553 to maybe answer your question ☺️

lwjohnst86 commented on docs/design/naming.qmd:
```diff
@@ -5,74 +5,137 @@ title: "Naming scheme"
+
+### App page URLs
+
+For the URL/app pages naming scheme, they follow a similar style to the
```
I'll split these out into other docs, so they are more of an explanation of the design, rather than a naming scheme.

lwjohnst86 commented on docs/design/naming.qmd:
```diff
@@ -5,74 +5,137 @@ title: "Naming scheme"
-projects <id> metadata create
-# This creates an associated data object at:
-# projects <id> metadata <id> data
+`packages/<id>/resources/<id>/edit`
```
No, if they need to edit something, they can always re-upload the same person's IDs with the updated data values and it will update in the database.

signekb replied:
Ohhh, what if they have a dataset with multiple data files for the same individual then?

lwjohnst86 replied:
I'm not sure I understand what you mean? Maybe you can ask the question over in the #554 PR about the outputs, and maybe give an example of what you mean.

lwjohnst86 commented on docs/design/naming.qmd:
````diff
@@ -5,74 +5,137 @@ title: "Naming scheme"
-projects <id> metadata <id> data update
-projects <id> metadata <id> data delete
-```
+:   Add data to a specific resource of a specific package.
````
I think we provide the full spec, give defaults if the user doesn't give them. There is no need for the user to dive into the frictionless spec or add additional fields that aren't standard (or ones that we might add to). Limit the choices the user has to make, so we can guide them toward an easier and smoother experience with structuring data.

signekb replied:
That makes sense. Thanks for elaborating 💪

lwjohnst86 commented on docs/design/naming.qmd:
```diff
@@ -81,45 +144,46 @@ environment with Postgres installed, the original input from the user
 will be stored in the Postgres database as the below file path and can
```
Ah, I see I missed updating some things, thanks for the catch!

lwjohnst86 commented on docs/design/naming.qmd:
```diff
@@ -81,45 +144,46 @@ environment with Postgres installed, the original input from the user
 will be stored in the Postgres database as the below file path and can
 be exported into actual files if desired.
```
I need to update this paragraph, I had missed it.

lwjohnst86 committed:
docs: :memo: add suggestions from PR review

Co-authored-by: Signe Kirk Brødbæk <40836345+signekb@users.noreply.github.com>

lwjohnst86 committed:
Merge branch 'main' of https://github.com/seedcase-project/seedcase-sprout into docs/update-naming-scheme

signekb commented on docs/design/naming.qmd:
```diff
@@ -5,74 +5,137 @@ title: "Naming scheme"
-    update, and delete
+-   Types of core objects: packages, resources, properties (data package metadata), and data.
+-   Types of actions taken on those objects: create, view, edit (a
+    file), add (data in the package), list, and delete.
```
Ah, yes, I see! Just thought that it would be clearer to use the same actions for functions and URLs. But I see your point 👍

lwjohnst86 committed:
docs: :memo: major rewrite of the naming scheme based on feedback

lwjohnst86 commented:
Based on everyone's feedback and discussions, I've updated the naming scheme and split out more of the description and explanation of applying the naming scheme into their own files (see #554, #553, and #555)

martonvago commented on docs/design/naming.qmd:
````diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
-# A human readable, auto-generated 'metadata' file
-ROOT/projects/<id>/README.md
-```
+This document describes a general naming scheme style for Sprout that
````
```suggestion
This document describes a general naming scheme for Sprout that
```
Maybe "scheme" is enough?

martonvago commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+We'll compose names based on the objects we and our users interact with
+as well as the actions taken on those objects. Following along with the
+Data Package terminology, we simplify "data package" to "package" and
+"data resource" to "resource". We use words "property" or "properties"
```
```suggestion
"data resource" to "resource". We use the words "property" or "properties"
```

martonvago commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+as well as the actions taken on those objects. Following along with the
+Data Package terminology, we simplify "data package" to "package" and
+"data resource" to "resource". We use words "property" or "properties"
+to describe the metadata about the package or resource that is contained
+within the `datapackage.json` file that follows the Frictionless Data
```
```suggestion
to describe package and resource metadata contained
in the `datapackage.json` file that follows the Frictionless Data
```
Just avoiding an extra that-clause

martonvago commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+| Object | Description |
+|----------------------------|--------------------------------------------|
+| package(s) | A data package that contains a collection of related data resources and properties. |
+| resource(s) | A single data file within a package. |
```
Would the (resource) properties be considered part of the resource here?

lwjohnst86 replied:
```suggestion
| resource(s) | A single data file within a package, along with its properties. |
```

martonvago commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+| resources | The folder that contains all resource files in a package. |
+| raw | The folder that contains raw data files for a data resource. |
+| raw data file | A compressed raw data file format, such as a `.zip`, `.gzip`, or `.tar.gz` file. |
+| JSON | The main way we pass and save information around. |
```
```suggestion
| JSON | The main way we save and pass information around. |
```
Reads better this way but conceptually the other way around is more accurate.

martonvago commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+| edit | Edit an object, specifically the properties object. |
+| add | Add another object to an existing object, specifically adding data to a resource. |
+| write | Write an object to a file. |
+| copy | Copy a file from on location to another. |
```
```suggestion
| copy | Copy a file from one location to another. |
```

martonvago reviewed (changes requested):
Very nice!! ⭐

lwjohnst86 committed:
docs: :pencil2: small edits

Co-authored-by: martonvago <57952344+martonvago@users.noreply.github.com>

lwjohnst86 committed:
docs: :pencil2: small edit

signekb commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+as well as by the [Frictionless
+Framework](https://v4.framework.frictionlessdata.io/) and [Data
+Package](https://datapackage.org/) structure. This naming scheme *only*
+applies to file paths, data objects, functions, URLs, API endpoints, and
```
```suggestion
applies to data objects, functions, URLs, API endpoints, and
```

signekb commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+
+We'll compose names based on the objects we and our users interact with
+as well as the actions taken on those objects. Following along with the
+Data Package terminology, we simplify "data package" to "package" and
```
```suggestion
[Data Package](https://datapackage.org/) terminology, we simplify "data package" to "package" and
```

signekb commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+Data Package terminology, we simplify "data package" to "package" and
+"data resource" to "resource". We use the words "property" or "properties"
+to describe package and resource metadata contained
+in the `datapackage.json` file that follows the Frictionless Data
```
Add link to Frictionless Data?

lwjohnst86 replied:
Included the word Frictionless to the data package above.

signekb commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+| Object | Description |
+|----------------------------|--------------------------------------------|
+| package(s) | A data package that contains a collection of related data resources and properties. |
+| resource(s) | A single data file within a package, along with its properties. |
```
Technically it can be thought of as a single data file (the parquet) or multiple data files (the raw file(s)), right?

lwjohnst86 replied:
Umm, not sure what you mean about the second part, since you suggested adding the properties word here.

signekb commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+| properties or property | Metadata about a package or resource. |
+| file structure or structure | The file and folder structure of a package. |
+| path | A file path listing the location of folders or files that Sprout interacts with. |
+| template | A template file for specific objects, such as a template properties file. |
```
```suggestion
| template | A template file for specific objects, such as a template for a package or resource properties file. |
```

signekb commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+| file structure or structure | The file and folder structure of a package. |
+| path | A file path listing the location of folders or files that Sprout interacts with. |
+| template | A template file for specific objects, such as a template properties file. |
+| data | The data file within a resource. |
```
Meaning raw or db or parquet? Or all three? :)

lwjohnst86 replied:
```suggestion
| data | The data file within a resource. Within Sprout, this includes the table in the database, the Parquet file, as well as raw files. |
```

signekb commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+| resources | The folder that contains all resource files in a package. |
+| raw | The folder that contains raw data files for a data resource. |
+| raw data file | A compressed raw data file format, such as a `.zip`, `.gzip`, or `.tar.gz` file. |
+| JSON | The main way we save and pass information around. |
```
Does this only include properties or also other information?

lwjohnst86 replied:
```suggestion
| JSON | The main way we save and pass information around on the package's or resource's properties. |
```

signekb commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+| raw | The folder that contains raw data files for a data resource. |
+| raw data file | A compressed raw data file format, such as a `.zip`, `.gzip`, or `.tar.gz` file. |
+| JSON | The main way we save and pass information around. |
+| Parquet | A columnar, efficient storage file format that we use to save a data resource. |
```
```suggestion
| Parquet | A columnar, efficient storage file format that we use to save data within a resource. |
```

signekb reviewed (changes requested):
Very nice 🎉

lwjohnst86 commented on docs/design/naming.qmd:
```diff
@@ -4,122 +4,63 @@ title: "Naming scheme"
+| Object | Description |
+|----------------------------|--------------------------------------------|
+| package(s) | A data package that contains a collection of related data resources and properties. |
+| resource(s) | A single data file within a package, along with its properties. |
```
```suggestion
| resource(s) | A single data file within a package, along with its properties and associated *raw* data files. |
```

lwjohnst86 committed:
docs: :memo: apply suggestions from code review

Co-authored-by: Signe Kirk Brødbæk <40836345+signekb@users.noreply.github.com>

lwjohnst86 committed:
docs: :memo: apply suggestions from code review

K-Beicher reviewed (commented):
Seems good to me.

lwjohnst86 committed:
Merge branch 'main' of https://github.com/seedcase-project/seedcase-sprout into docs/update-naming-scheme
